### PR TITLE
docker_secret, docker_config: allow to base64-decode data

### LINKED ISSUE #35119: Creating Docker secrets from file system
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

docker_secret

##### ANSIBLE VERSION

```
ansible 2.4.0.0
  config file = None
  configured module search path = [u'/Users/me/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.4.0.0/libexec/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:54:19) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```

##### CONFIGURATION

NA
##### OS / ENVIRONMENT

N/A

##### SUMMARY

I think that docker_secret is an excellent addition to Ansible and I am interested to use it to manage infrastructure. I looked at it today but it appears that I can only create secrets from strings, not directly from files (e.g. I have a property file that I would like to turn into a secret or convert an SSL private key into a secret). I considered `cat`ing the file into a variable but I was hoping for something more direct. Is this correct and I'm limited to only strings? I understand it's only in preview so it's not the final iteration. If I can't create secrets from files, I'd be glad to add functionality to the module with some guidance. 

##### STEPS TO REPRODUCE

I was hoping to do something like the below.
```
name: create a new docker secret
docker_secret:
  name: my_new_secret
  data: /path/to/secret
  file: yes
  state: present
```

##### ACTUAL RESULTS

```

```

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
This allows to pass binary data or JSON data Base64-encoded, so that Ansible won't destroy / reformat it. Fixes #35119.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docker_secret
docker_config

### CODE REVIEW

**Comment by felixfontein** [comment]
(docker_config will first appear in 2.8, so there's no changelog for the modifications to it.)

**Comment by felixfontein** [comment]
Rebased to resolve conflict.

ready_for_review

File: lib/ansible/modules/cloud/docker/docker_secret.py [hunk 1]
```diff
@@ -72,10 +84,14 @@
 
 EXAMPLES = '''
 
-- name: Create secret foo
-  docker_secret:
+- name: Create secret foo (from a file on the control machine)
+  docker_config:
```
Comment (DBendit) [inline lib/ansible/modules/cloud/docker/docker_secret.py]: This seems like an accidental copy/paste

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_secret.py]: Indeed! Thanks for spotting this!

**Comment by felixfontein** [comment]
@ushuz while you're around, can you take a look here? :)

**Comment by ushuz** [comment]
@felixfontein The code looks good to me. 

However, I still hold my opinion here https://github.com/ansible/ansible/issues/35119#issuecomment-442805233, that this should be taken care at a higher level. 

If urgent, this is a good workaround and can help some people. If it's not that urgent, I think we could take time, as this may set a precedence for other modules to follow. But I may be worrying too much, your call.

**Comment by felixfontein** [comment]
I'm afraid there's no way this will be solved at a higher level for at least some Ansible versions... And for binary data, it could be that there's never a solution. And other modules (such as `slurp`) have been using Base64 encoding for a long time already (for that reason).

**Comment by ushuz** [comment]
shipit

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Thanks!

**Comment by felixfontein** [comment]
@DBendit @ushuz thanks for your reviews!
@gundalow thanks for merging!

**Pull request merged**